components/event/org.wso2.carbon.event.core/src/main/java/org/wso2/carbon/event/core/internal/topic/registry/RegistryTopicManager.java

```
@@ -398,19 +482,96 @@ public boolean removeTopic(String topicName) throws EventBrokerException {
+     */
+    private static void authorizePermissionsToLoggedInUser(String username, String destinationName,
+                                                           String destinationId,
+                                                           UserRealm userRealm) throws
+                                                                                UserStoreException {
+
+        //For registry we use a modified queue name
+        String newDestinationName = destinationName.replace("@", AT_REPLACE_CHAR);
+
+        // creating the internal role name
+        String roleName = UserCoreUtil.addInternalDomainName(TOPIC_ROLE_PREFIX +
+                                                             newDestinationName.replace("/", "-"));
+
+        // the interface to store user data
+        UserStoreManager userStoreManager = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getUserStoreManager();
+
+        if (!userStoreManager.isExistingRole(roleName)) {
+            String[] user = {MultitenantUtils.getTenantAwareUsername(username)};
+
+            // adds the internal role to user store
+            userStoreManager.addRole(roleName, user, null);
+            // gives subscribe permissions to the internal role in the user store
+            userRealm.getAuthorizationManager().authorizeRole(
+                    roleName, destinationId, EventBrokerConstants.EB_PERMISSION_SUBSCRIBE);
+            // gives publish permissions to the internal role in the user store
+            userRealm.getAuthorizationManager().authorizeRole(
+                    roleName, destinationId, EventBrokerConstants.EB_PERMISSION_PUBLISH);
+            // gives change permissions to the internal role in the user store
+            userRealm.getAuthorizationManager().authorizeRole(
+                    roleName, destinationId, EventBrokerConstants.EB_PERMISSION_CHANGE_PERMISSION);
+
+        } else {
+            log.warn("Unable to provide permissions to the user, " +
```

Should we throw an exception here?